### DISCUSSION TITLE
Disable throttling programmatically

### DISCUSSION THREAD

**Discussion by EvilKarter** [GoogleChrome/lighthouse#15723] [Q&A] [answered]
Hi I am trying to use lighthouse together with puppeteer. I want to deactivate throttling completely. 

It is explained how this works on the [CLI](https://github.com/GoogleChrome/lighthouse/blob/main/readme.md#cli-options) but I have not found anything to solve this differently in Node. So I have adopted the solution from the CLI. 

However, the output of the configSettings in the result confuses me. 
The throttling is still there. Is this normal or am I doing something wrong?

Here my test file
```javascript
(async () => {
  const puppeteer = await import('puppeteer');
  const lighthouse = await import('lighthouse');
  // Use Puppeteer to launch headless Chrome see https://github.com/GoogleChrome/lighthouse/blob/main/docs/puppeteer.md
  // - Omit `--enable-automation` (See https://github.com/GoogleChrome/lighthouse/issues/12988)
  // - Don't use 800x600 default viewport
  const browser = await puppeteer.launch({
    // Set to false if you want to see the script in action.
    headless: 'new',
    defaultViewport: null,
    ignoreDefaultArgs: ['--enable-automation']
  });

  try {
    const page = await browser.newPage();
    await page.setViewport({
      width: 375,
      height: 800,
      deviceScaleFactor: 2,
      isMobile: true,
    });
    await page.setUserAgent('My Custom User Agent');
    const url = 'https://www.google.com';

    await page.goto(url);
    const lighthouseFlags = {
      port: new URL(browser.wsEndpoint()).port,
    };

    const lighthouseConfig = {
      // extends property only supports extension of lighthouse:default see https://github.com/GoogleChrome/lighthouse/blob/main/docs/configuration.md#config-extension
      extends: 'lighthouse:default',
      settings: {
        output: 'json',
        locale: 'en-US',
        onlyCategories: ['accessibility'],
        disableFullPageScreenshot: true,
        // Disable device emulation and all throttling after example from see https://github.com/GoogleChrome/lighthouse/blob/main/readme.md#cli-options
        throttlingMethod: 'provided',
        screenEmulation: {
          disabled: true
        },
        emulatedUserAgent: false,
        // see https://github.com/GoogleChrome/lighthouse/blob/main/docs/configuration.md#more-examples
        skipAudits: [
          // There are always bf-cache failures when testing in headless. Reenable when headless can give us realistic bf-cache insights.
          'bf-cache',
        ],
      },
    };
    const runnerResult = await lighthouse.default(url, lighthouseFlags, lighthouseConfig, page);
    process.stdout.write(JSON.stringify(runnerResult.lhr.configSettings));
  } catch (e) {
    process.stdout.write(e.message);
  }
  await browser.close();
})();
``` 

here the config Settings output

```javascript
{
  "output": "json",
  "maxWaitForFcp": 30000,
  "maxWaitForLoad": 45000,
  "pauseAfterFcpMs": 5250,
  "pauseAfterLoadMs": 5250,
  "networkQuietThresholdMs": 5250,
  "cpuQuietThresholdMs": 5250,
  "formFactor": "mobile",
  "throttling": {
    "rttMs": 150,
    "throughputKbps": 1638.4,
    "requestLatencyMs": 562.5,
    "downloadThroughputKbps": 1474.5600000000002,
    "uploadThroughputKbps": 675,
    "cpuSlowdownMultiplier": 4
  },
  "throttlingMethod": "provided",
  "screenEmulation": {
    "mobile": true,
    "width": 412,
    "height": 823,
    "deviceScaleFactor": 1.75,
    "disabled": true
  },
  "emulatedUserAgent": false,
  "auditMode": false,
  "gatherMode": false,
  "clearStorageTypes": [
    "file_systems",
    "shader_cache",
    "service_workers",
    "cache_storage"
  ],
  "disableStorageReset": false,
  "debugNavigation": false,
  "channel": "node",
  "usePassiveGathering": false,
  "disableFullPageScreenshot": true,
  "skipAboutBlank": false,
  "blankPage": "about:  blank",
  "ignoreStatusCode": false,
  "budgets": null,
  "locale": "en-US",
  "blockedUrlPatterns": null,
  "additionalTraceCategories": null,
  "extraHeaders": null,
  "precomputedLanternData": null,
  "onlyAudits": null,
  "onlyCategories": [
    "accessibility"
  ],
  "skipAudits": [
    "bf-cache"
  ]
}

```

**Comment by adamraine** [comment]
`throttlingMethod: 'provided'` should remove all throttling like you want.

What do you mean by "The throttling is still there"?

**Reply by EvilKarter** [reply]
Hi adamraine, 
I am wondering why in the configSettings section of the LHR the throttling is still listed with details.
```javascript
"throttling": {
    "rttMs": 150,
    "throughputKbps": 1638.4,
    "requestLatencyMs": 562.5,
    "downloadThroughputKbps": 1474.5600000000002,
    "uploadThroughputKbps": 675,
    "cpuSlowdownMultiplier": 4
  },
  "throttlingMethod": "provided",
``` 
Maybe I'm missing something, but how can I tell from the result alone whether the throttling was provided with or without further details if the throttling is always output in the result? 

Because if I pass throttlingMethod provided and also set the rttMs to 10, I see the same thing in the result except for the rttMs value I set. 
test script: 
```javascript 
 throttlingMethod: 'provided',
 throttling: {
    "rttMs": 10
 },
``` 
result: 
```javascript
"throttling": {
    "rttMs": 10,
    "throughputKbps": 1638.4,
    "requestLatencyMs": 562.5,
    "downloadThroughputKbps": 1474.5600000000002,
    "uploadThroughputKbps": 675,
    "cpuSlowdownMultiplier": 4
  },
  "throttlingMethod": "provided",
```

**Reply by adamraine** [reply, marked as answer]
The parameters are ignored by Lighthouse when `throttlingMethod: 'provided'` is set. They still appear in the JSON and you can adjust there value, but they don't affect the page load.

**Reply by SivasubramanianV** [reply]
@adamraine which throttling replicates a real world performance view. Provided or simulated or packet? And also can you please explain why. Also what kind of cpu throttling should we set?

**Reply by adamraine** [reply]
@SivasubramanianV please see our [throttling documentation](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md) the trade-offs of different throttling methods.

**Reply by SivasubramanianV** [reply]
@adamraine thanks..i read it but there is no mention of 'provided' setting. It only explains packet setting which is a particular machines setting-not sure how that could be an indicator of performance overall as network speed and device speed may vary widely. Also I would like to understand which is ideal as its tough to decide between provided and simulated. So far we have been using simulated and the scores are low so my team is planning to use provided, would that be a right approach. Or i would atleast like to know across industry what is used as a practice.

**Reply by adamraine** [reply]
> Also I would like to understand which is ideal as its tough to decide between provided and simulated. So far we have been using simulated and the scores are low so my team is planning to use provided, would that be a right approach.

`provided` just means Lighthouse will not add any throttling to the page automatically. Lighthouse will just inherit the network and CPU conditions and of your system, so if there is throttling set up system-wide then it will still affect the Lighthouse result. You should only use `provided` throttling if you plan to manually set up the throttling environment yourself.

> Also I would like to understand which is ideal as there is no clear better method suggested.

There isn't really a a single "ideal" approach, but packet-level throttling tends to be the most accurate. The trade-off is that it can introduce more variance and typically needs to be set up at the system level.

**Reply by SivasubramanianV** [reply]
@adamraine thanks.
So for packet level, what is the throttling-method? Is it devtools? I saw the documentation but just wanted to confirm.

Also if i use 'provided', any idea how to setup system level throttling myself?

**Reply by adamraine** [reply]
> So for packet level, what is the throttling-method? Is it devtools? I saw the documentation but just wanted to confirm.

[This example](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md#using-lighthouse-with-sitespeediothrottle) will disable network throttling but keep CPU throttling. `--throttling-method=provided` would disable both CPU and network throttling.

> Also if i use 'provided', any idea how to setup system level throttling myself?

There are many different libraries that do it, here is one example from our docs:

https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md#how-do-i-get-packet-level-throttling

**Reply by SivasubramanianV** [reply]
> > Also I would like to understand which is ideal as its tough to decide between provided and simulated. So far we have been using simulated and the scores are low so my team is planning to use provided, would that be a right approach.
> 
> `provided` just means Lighthouse will not add any throttling to the page automatically. Lighthouse will just inherit the network and CPU conditions and of your system, so if there is throttling set up system-wide then it will still affect the Lighthouse result. You should only use `provided` throttling if you plan to manually set up the throttling environment yourself.
> 
@adamraine is it that i should only use 'provided' if i am setting up throttling environment myself. Can i not just give 'provided' without setting up throttling environment and measure the lighthouse score. Trying to understand if there is anything wrong in that way of measurement?

**Reply by adamraine** [reply]
The goal of throttling is to create an network and CPU environment similar to what real users are experiencing. If you just use `provided` without setting up any additional throttling on your system then you will be seeing the performance metrics for your local workstation. Unless your users tend to have similar network connections and hardware, these numbers won't be very useful.

**Reply by SivasubramanianV** [reply]
@adamraine thanks, i use https://www.npmjs.com/package/@sitespeed.io/throttle for network throttling in provided, is there something that i can use for CPU throttling. @adamraine any suggestions for this?
